ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+                indexSegment.getStartOffset())) {
+              validEntries.add(entry);
+            }
+          } else {
+            validEntries.add(entry);
+          }
+        } else {
+          if (srcIndex.isExpired(currentFinalState)) {
+            logger.trace("{} in index segment with start offset {} in {} is not valid because it is an expired PUT",
+                entry, indexSegment.getStartOffset(), storeId);
+            continue;
+          }
+          if (currentFinalState.isPut()) {
+            if (currentFinalState.getLifeVersion() != currentValue.getLifeVersion()) {
+              throw new IllegalStateException(
+                  "Two different lifeVersions  for puts key" + currentKey + " in store " + dataDir);
+            }
+            validEntries.add(entry);
+          } else if (currentFinalState.isDelete()) {
+            if (currentFinalState.getOperationTimeInMs() >= compactionLog.getCompactionDetails().getReferenceTimeMs()) {
+              validEntries.add(entry);
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  entry, indexSegment.getStartOffset(), storeId);
+            }
+          } else {
+            validEntries.add(entry);
+          }
+        }
+      }
+      return validEntries;
+    }
+
+    /**
+     * Determines whether a TTL update entry is valid. A TTL update entry is valid as long as the associated PUT record
+     * is still present in the store (the validity of the PUT record does not matter - only its presence/absence does).
+     * @param key the {@link StoreKey} being examined
+     * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} that the TTL update record is in
+     * @return {@code true} if the TTL update entry is valid
+     * @throws StoreException if there are problems reading the index
+     */
+    private boolean isTtlUpdateEntryValidWhenFinalStateIsDeleteAndRetention(StoreKey key,
+        Offset indexSegmentStartOffset) throws StoreException {
+      boolean valid = false;
+      //  A TTL update entry is "valid" if the corresponding PUT is still alive
+      // The PUT entry, if it exists, must be "before" this TTL update entry.
+      FileSpan srcSearchSpan = new FileSpan(srcIndex.getStartOffset(), indexSegmentStartOffset);
+      IndexValue srcValue = srcIndex.findKey(key, srcSearchSpan, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+      if (srcValue == null) {
+        // PUT is not in the source - it might be compacted
+        logger.trace("TTL update of {} in segment with start offset {} in {} is not valid the corresponding PUT entry "
+            + "does not exist anymore", key, indexSegmentStartOffset, storeId);
+      } else {
+        // Exists in the source. This can happen either because
+        // 1. The FileSpan to which srcValue belongs is not under compaction (so there is no reason for tgt to have it)
```

Comment:
To make the comment clearer. could you change `srcValue` to `PUT`?
(same for `it`, if I understand correctly, `it` is also referring to `PUT`)

Reply:
sure